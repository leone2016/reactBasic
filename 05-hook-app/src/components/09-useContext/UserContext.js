import {createContext} from 'react';

/**
 * UserContext es un high order component, quiere decir todo lo que tenga componentes dentro, ejemplo 
 * en AppRouter.js el <Router> contiene al <Switch/> y este a su vez tiene al <Route>, por ende, Route y Switch 
 * sin high order component
 *  
 * Se le puede asignar componentes hijos dentro de el
 */
export const UserContext = createContext(null)