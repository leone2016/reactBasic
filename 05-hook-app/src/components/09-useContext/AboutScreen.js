import React, { useContext } from 'react'
import { UserContext } from './UserContext';

export const AboutScreen = () => {
    const {setUser} = useContext(UserContext);
    return (
        <div>
            <h2>About Screen</h2>
            <hr/>   

        </div>
    )
}
