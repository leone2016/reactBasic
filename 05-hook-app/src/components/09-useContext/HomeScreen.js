import React, { useContext } from 'react'
import { UserContext } from './UserContext';

export const HomeScreen = () => {
    const userContext = useContext(UserContext);
    console.log(userContext);
    return (
        <div>
            <h2>Home Screen</h2>
            <hr/>
        </div>
    )
}
