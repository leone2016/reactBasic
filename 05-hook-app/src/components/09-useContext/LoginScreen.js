import React, { useContext } from 'react'
import { UserContext } from './UserContext';

export const LoginScreen = () => {
    const {setUser} = useContext(UserContext);
    return (
        <div>
            <h2>Login Screen</h2>
            <hr/>
            <button className="btn btn-primary" onClick={()=> setUser({
                id: 1991,
                name: 'Patricio',
                email: 'leoz.31@hotmail.com',
            })}> LOGIN </button>
        </div>
    )
}
