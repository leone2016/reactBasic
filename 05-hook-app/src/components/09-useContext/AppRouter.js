import React from 'react'
import {
    BrowserRouter as Router,
    Switch,
    Route
} from "react-router-dom";
import { AboutScreen } from "./AboutScreen.js"
import { LoginScreen } from "./LoginScreen.js"
import { HomeScreen } from "./HomeScreen.js"
import { NavBar } from './NavBar.js';
export const AppRouter = () => {
    return (
        <Router>
            <div>
                <NavBar />
                {/*tienen la misma funcion del switch de Js, dependiendo de la funcion muestra algo */}
                <Switch>
                    <Route exact path="/" component={HomeScreen} />
                    <Route exact path="/about" component={AboutScreen} /> 
                    <Route exact path="/login" component={LoginScreen} />
                    <Route component={HomeScreen} /> {/* path por defecto, se puede mostrar 404 */}
                </Switch>
            </div>
        </Router>
    )
}
