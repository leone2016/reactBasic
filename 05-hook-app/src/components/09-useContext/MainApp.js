import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'

export const MainApp = () => {

    const [user, setUser] = useState({});

    return (
        <div>
            {/*   <h2>MAin Screen</h2>
            <hr/> */}
            <UserContext.Provider value={{
                user,
                setUser,
                temporal: 1234
            }}>
                <AppRouter />
            </UserContext.Provider>

        </div>
    )
}
