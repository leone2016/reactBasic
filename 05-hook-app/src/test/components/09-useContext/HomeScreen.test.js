import React from 'react';
import { shallow, mount} from 'enzyme';
import { HomeScreen } from '../../../components/09-useContext/HomeScreen';
import { UserContext } from '../../../components/09-useContext/UserContext';
// clase 158
describe('Pruebas en <HomeScreen>', () => {

    const user = {
        name: 'leonardo',
        email: 'leoz.31@hotmail.com'
    }
    // es es necesario definir el context, ya que el contexto necesito un usuario
    /**
     * da el siguiente error:: TypeErrror: Cannot destructure property `user` of `undefined` or `null`
     * 1. primero se crea el contexto <UserContext>
     * 2. se usa el contexto en MainApp.js con el cual se crea un value que va compartir a lo largo con todos sus componente
     *      con todos sus componentes hijos en este caso es el <AppRouter/>
     */

     /**
      * Aqui se puede ver la diferencias de  shallow y mount
      * 
      * shallow usa high order component, eso quiere decir que solo haya snapshoot de   <UserContext/>
      * y no de <HomeScreen>, para ello se utiliza mount, cuando se necesita renderizar todo en general
      */
    const wrapper = mount(
        <UserContext.Provider value={{
            user
        }}>
            <HomeScreen />
        </UserContext.Provider>

    );
    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    })
});
