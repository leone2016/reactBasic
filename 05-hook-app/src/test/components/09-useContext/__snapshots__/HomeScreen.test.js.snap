// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HomeScreen> debe de mostrarse correctamente 1`] = `
<div>
  <h2>
    Home Screen
  </h2>
  <hr />
  <pre>
    {
    "name": "leonardo",
    "email": "leoz.31@hotmail.com"
}
  </pre>
</div>
`;
